Store: Support setup of stores in any country; redirect to Woo wizard when needed

allendav opened the pull request:
Finishes #20877 

This code:
* changes the setup flow AddressView to use showAllLocations mode (thereby opening it up to all countries)
* tweaks `lib/countries` to NOT provide defaults for currency, dimensions, and weight for nonUS nonCA countries (thereby allowing the Woo wizard to do it itself)
* redirects the user to the Woo wizard if they provide a nonUS nonCA address in the "Howdy! Ready to start selling? First we need to know where you are in the world." view
* adds a `QueryLocations` component to kick off fetching locations
* adds a `getCountriesWithStates` selector so `StoreLocationSetupView` can appropriately require (or not) a state to be selected

Notes:
* I want to (later) refactor `AddressView` to a `connect`-ed component so it can manage things like whether or not a country has states

To test:
* `npm run test-client client/extensions/woocommerce/`
* Use https://developer.wordpress.com/docs/api/console/ to set set_store_address_during_initial_setup to 0 for the site using the POST /sites/{siteID}/calypso-preferences/woocommerce endpoint
* Navigate to http://calypso.localhost:3000/store/{domain}
* Test each of the following addresses. After hitting let's go, ensure 1) that you are dropped on your store's CALYPSO dashboard and 2) that the currency, dimensions and weight are set appropriately in settings-payments and product-add (repeat the developer console business above to regain access to the form each time needed)

```
15517 Main St
Mill Creek WA 98012
United States
(Make sure currency is set to USD, dimensions to in and weight to lbs)

811 Hornby St
Vancouver, BC V6Z2E6
Canada
(Make sure currency is set to CAD, dimensions to cm and weight to kg)
```

* Next, test each of the following addresses. After hitting let's go, ensure 1) that you are dropped on your store's WP-ADMIN WOOCOMMERCE SETUP WIZARD and 2) that the address has been copied over for you.
* Make sure for countries that have states, that you are required to choose one before you are allowed to submit the form.  For countries that do not, make sure you are not required to choose one :)
* NOTE: We do NOT push a default for nonUS nonCA stores at this time for currency, dimensions or weight. Your previous settings on the site will be retained, or, in the event this is a brand new site, WooCommerce defaults for that country will be automatically used in the wizard.

```
2 Elizabeth St
Melbourne VIC 3004
Australia

7 Allées du Président Franklin Roosevelt
31000 Toulouse
France
```

timmyc commented on client/extensions/woocommerce/app/dashboard/store-location-setup-view.js:
```diff
@@ -134,6 +129,15 @@ class StoreLocationSetupView extends Component {
+			// If we don't support a calypso experience yet for this country, let
+			// them complete setup with the wp-admin WooCommerce wizard
+			if ( ! isStoreManagementSupportedInCalypsoForCountry( this.state.address.country ) ) {
+				const storeSetupURL =
```
This might be a good thing to add some `tracks` around right here. Or an mc stat bump at least.

allendav replied:
Yeah, agreed!

timmyc commented on client/extensions/woocommerce/app/dashboard/index.js:
```diff
@@ -132,6 +133,8 @@ class Dashboard extends Component {
 			setupChoicesLoading,
 		} = this.props;
 
+		const adminURL = get( selectedSite, 'options.admin_url', '' );
```
Is there any chance we will indeed get the default empty string value here? If so should we ensure we have a string with length prior to attempting to window.location = ''?

allendav replied:
I don't think so, but we could test for that and console an error perhaps?

allendav replied:
Added `warn` in https://github.com/Automattic/wp-calypso/pull/21297/commits/79f4d91320a4f456fe0d4f7a2400fb40395ff473

timmyc commented:
Testing nicely and everything working as expected. As I noted in our DM convo - the standard dashboard is briefly shown/flashed prior to `window.location` kicking in. Might be nice to find a way to avoid that here, and like you said we will likely need a solution to redirect non-supported countries that attempt to access https://wordpress.com/store/domain to their wp-admin.

allendav commented:
@timmyc wrote:

> and like you said we will likely need a solution to redirect non-supported countries that attempt to access https://wordpress.com/store/domain to their wp-admin.

Agreed - opened https://github.com/Automattic/wp-calypso/issues/21304 for that

allendav commented:
@timmyc wrote:

> As I noted in our DM convo - the standard dashboard is briefly shown/flashed prior to window.location kicking in. Might be nice to find a way to avoid that here

Fixed in https://github.com/Automattic/wp-calypso/pull/21297/commits/7a75ccadbe1b3c005a1238057d6548c722ed618b and also fixed a glitch where we weren't pre-filling with nonUS nonCA addresses from domain registration

timmyc commented:
Flash is gone! As I mentioned in Slack, I'd like to give this some more thorough testing on Monday doing the full signup flow... maybe toss in a IE experience in there too just for kicks... and maybe have one more set of 👀 on it too. But so very super stoked for how far you've gotten us on this project today 🥇

allendav commented:
@timmyc : I added bumping a stat during setup for the store country in https://github.com/Automattic/wp-calypso/pull/21297/commits/60ae075be89361f2f8dfeda5725f0e2e544576d2

calypso_woo_store_setup_country will be bumped with the country code

justinshreve commented on client/extensions/woocommerce/app/dashboard/store-location-setup-view.js:
```diff
@@ -148,6 +141,18 @@ class StoreLocationSetupView extends Component {
 			// before the next step's dialog displays
+
+			// mc stat 32 char max :P
+			analytics.mc.bumpStat( 'calypso_woo_store_setup_country', this.state.address.country );
```
Just a note that I logged an issue at https://github.com/Automattic/wp-calypso/issues/21336, so that later we can update this to use a method that ignores test sites.

gwwar replied:
Please avoid using bare analytics lib calls. Testing becomes much more difficult, since analytics assumes window is available. We do have analytics middleware available, which allows us to dispatch a redux action, or add additional meta data so we can fire this off. See: https://github.com/Automattic/wp-calypso/tree/master/client/state/analytics and let us know if you need help walking through that

justinshreve commented on client/extensions/woocommerce/state/sites/settings/actions.js:
```diff
@@ -80,53 +79,69 @@ export const doInitialSetup = (
+	] );
+
 	if ( pushDefaultsForCountry ) {
+		// TODO Support other currency positions, post-v1 etc. See https://github.com/Automattic/wp-calypso/issues/15498
```
It might be worth a comment like

> in the event this is a brand new site, WooCommerce defaults for that country will be automatically used in the wizard.

so it makes it clear that other countries will still get their correct values though the wizard.

allendav replied:
Fixed in https://github.com/Automattic/wp-calypso/pull/21297/commits/20e50be7b870bb2e5ff1c9223c1c55de64434163

allendav commented:
@justinshreve wrote:

> Should we change the language on the first screen now?
> If you're in the United States or Canada, you can sell your products right on your site and ship them to customers in a snap!

Agreed. I'll change it to simply "You can sell your products right on your site and ship them to customers in a snap!"

kellychoffman commented:
Looking good!

Feedback on the non supported country scenarios: 

1. Could we use the `busy` state on the 'Let's go' button once its pushed? There's a bit of a waiting period before the wizard is shown and it would be good to indicate that things are indeed happening. https://wpcalypso.wordpress.com/devdocs/design/buttons 

2. I clicked the browser back button and got the forever loading "Finishing setup" screen. Should we redirect `/store/URL` to `URL/wp-admin`? 
![screen shot 2018-01-08 at 3 44 09 pm](https://user-images.githubusercontent.com/5835847/34696999-39e45dd6-f48e-11e7-843c-a6d69c02bb34.png)

3. The wizard also uses "Let's go!" for the button name in the first step. Should we change that to something else since they already pressed "Let's go!" in the step prior? Perhaps just "Next"? 

3. The transition from the address step to the WooCommerce wizard is pretty abrupt. I think we should mention WooCommerce in the address step if the user picks a non supported country to ease that transition. Will whip up a quick mock. (This could go in another PR if its out of scope)

allendav commented:
@kellychoffman wrote

> Could we use the busy state on the 'Let's go' button once its pushed? There's a bit of a waiting period before the wizard is shown and it would be good to indicate that things are indeed happening

Fixed in https://github.com/Automattic/wp-calypso/pull/21297/commits/89af08694d4b7e7850b7f9ff9c4f95d326dd3102

allendav commented:
@kellychoffman wrote:

> I clicked the browser back button and got the forever loading "Finishing setup" screen. Should we redirect /store/URL to URL/wp-admin?

We have https://github.com/Automattic/wp-calypso/issues/21304 coming up for that - is that ok?

allendav commented:
@kellychoffman wrote:

> The wizard also uses "Let's go!" for the button name in the first step. Should we change that to something else since they already pressed "Let's go!" in the step prior? Perhaps just "Next"?

Fixed in https://github.com/Automattic/wp-calypso/pull/21297/commits/022c1ace80739650335dfc4c062790b27dfaee2d

allendav commented:
@kellychoffman wrote

> The transition from the address step to the WooCommerce wizard is pretty abrupt. I think we should mention WooCommerce in the address step if the user picks a non supported country to ease that transition.

Let's do it in a new issue : https://github.com/Automattic/wp-calypso/issues/21376

a8ci18n commented on client/extensions/woocommerce/app/dashboard/store-location-setup-view.js:
```diff
@@ -215,11 +228,13 @@ class StoreLocationSetupView extends Component {
 					disabled={ submitDisabled }
 					onClick={ this.onNext }
-					label={ translate( "Let's Go!" ) }
+					label={ translate( 'Next', { context: 'Label for button that submits a form' } ) }
```
Hi! I've found a possible matching string that has already been translated **86** times:
`translate( 'Next', { context: 'Button label for a next revision'} )` *ES Score: 12*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/fix/20877-redirect-non-us-non-ca-woo-wizard)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

:information_source: This **context** is really long. Are you sure you don't want to use a translator comment instead?

kellychoffman commented:
> We have #21304 coming up for that - is that ok?

👍

gwwar commented on client/extensions/woocommerce/components/query-locations/index.js:
```diff
@@ -0,0 +1,65 @@
+		}
+	};
+
+	componentWillReceiveProps = ( { siteId, loaded } ) => {
```
Side effects should be performed in `componentDidMount` and `componentDidUpdate`. See this [cheat sheet](https://gist.github.com/bvaughn/923dffb2cd9504ee440791fade8db5f9)

allendav replied:
Thanks @gwwar - will fix (and open issues for other places we are doing this in woocommerce)

allendav replied:
Fixed in https://github.com/Automattic/wp-calypso/pull/21297/commits/698f0228879b86ac921f12fd74df554610db7238

gwwar commented on client/extensions/woocommerce/components/query-locations/index.js:
```diff
@@ -0,0 +1,65 @@
+	componentWillMount = () => {
+		const { siteId, loaded } = this.props;
+
+		if ( siteId && ! loaded ) {
```
Is this checking for fetching state, or if something exists? If it's the latter you likely don't want it since we'll end up with stale data if a user revisits the same view without reloading the page.

allendav replied:
Its the former (was state fetched). That said, the data _can_ vary by site, but is highly unlikely to change on a given site (requires a plugin update to change.)

gwwar replied:
So with a QueryComponent the general pattern is:
- Let the component render with current available data (which may be stale)
- But also fetch new data in 2 cases:
-- On Mounts
-- Or when Props change

The problem with not fetching when some state is loaded, is that we won't will fetch again until the user refreshes the page. (Or in the case of us persisting state, never again)

allendav replied:
Fixed in https://github.com/Automattic/wp-calypso/pull/21297/commits/698f0228879b86ac921f12fd74df554610db7238

gwwar commented on client/extensions/woocommerce/components/query-locations/index.js:
```diff
@@ -0,0 +1,65 @@
+			fetchLocations,
+		},
+		dispatch
+	);
```
If you aren't doing anything fancy, use the object shorthand:
```jsx
const mapDispatchToProps = { fetchLocations };
```

allendav replied:
Fixed in https://github.com/Automattic/wp-calypso/pull/21297/commits/698f0228879b86ac921f12fd74df554610db7238

allendav commented:
@justinshreve @timmyc @gwwar since circleci hates this branch, let's wrap this up over here: https://github.com/Automattic/wp-calypso/pull/21409

timmyc commented:
Should this PR be closed out then @allendav?

allendav commented:
@timmyc - i've removed the needs review label for the moment, but lets leave it open for a little while to avoid losing the circleci situation

allendav commented:
Giving up on this branch. We delivered from #21409 instead. Never did resolve why circleci hated this branch
